beetsplug/discogs.py

```diff
@@ -180,6 +180,62 @@ def candidates(self, items, artist, album, va_likely, extra_tags=None):
             self._log.debug('Connection error in album search', exc_info=True)
             return []
 
+    def item_candidates(self, item, artist, title):
+        """Returns a list of TrackInfo objects for Search API results
+        matching ``title`` and ``artist``.
+        :param item: Singleton item to be matched.
+        :type item: beets.library.Item
+        :param artist: The artist of the track to be matched.
+        :type artist: str
+        :param title: The title of the track to be matched.
+        :type title: str
+        :return: Candidate TrackInfo objects.
+        :rtype: list[beets.autotag.hooks.TrackInfo]
+        """
+        if not self.discogs_client:
+            return
+
+        if not artist and not title:
+            self._log.debug('Skipping Discogs query. File missing artist and '
+                            'title tags.')
+
+        query = f'{artist} {title}'
+        try:
+            albums = self.get_albums(query)
+        except DiscogsAPIError as e:
+            self._log.debug('API Error: {0} (query: {1})', e, query)
+            if e.status_code == 401:
+                self.reset_auth()
+                return self.item_candidates(item, artist, title)
+            else:
+                return []
+        except CONNECTION_ERRORS:
+            self._log.debug('Connection error in track search', exc_info=True)
+        candidates = []
+        for album_cur in albums:
+            self._log.debug(u'searching within album {0}', album_cur.album)
+            track_list = self.get_tracks_from_album(album_cur)
+            candidates += track_list
+        for candidate in candidates:
+            candidate.data_source = 'Discogs'
+        # first 10 results, don't overwhelm with options
+        return candidates[:10] 
+
+    def get_tracks_from_album(self, album_info):
+        """Return a list of tracks in the release
+        """
+        if not album_info:
+            return []
+
+        result = []
+        for track_info in album_info.tracks:
+            # attach artist info if not provided
+            if not track_info['artist']:
+                track_info['artist'] = album_info.artist
+                track_info['artist_id'] = album_info.artist_id
```

Comment:
I played around a little and found out that simply applying data_url from the album here would do the trick:
```
            track_info['data_url'] = album_info.data_url
```

I think the `if not` is not required since data_url will probably always be required.

Looks like that then :-)

```
/remote/data/music-library/DrumnbassJungleDrumfunk/0-Album-Missing/Crossfire - Firebolt.mp3
Tagging track: Crossfire - Firebolt
URL:
    https://www.discogs.com/release/125946-Crossfire-Counterforce-Firebolt
(Similarity: 93.3%) (source) (Discogs)
[A]pply, More candidates, Skip, Use as-is, Enter search, enter Id, aBort,
eDit, edit Candidates, plaY?
```

Reply:
Thanks for digging into this! I'll make that edit.

I don't think that `if not` hurts anything (besides performance negligibly). The real question for me is: if there is a URL there already would it be something reasonable and would we want to pass it along, or is the URL we're providing via `album_info.data_url` always the "right one"?
For me the answer to this question would determine whether to use the `if not`.

Reply:
I'm pretty sure album_info.data_url is the _only_ one we'll get, haha ;-) Certainly you are right, the `if not` doesn't really hurt. I just though that it would even be cleaner and "kind of a statement" to say here: There is no track_url, because it's just not existing on discogs, thus we simply apply the album_url to the track so the importer prompt shows it similar to the track_urls musicbrainz and other metadata source plugins do it.

Here some debugging around these lines: FIXME

```
$ beet import -I -t -s /remote/data/music-archive/DrumnbassJungleDrumfunk/0-Single-Tracks/Crossfire/Crossfire\ -\ Hideout-sour.mp3 

This is data_url in get_album_info:

https://www.discogs.com/release/76357-Crossfire-Hideout-Swordfight



An this is the whole result.data object in get_album_info:

{'artists': [{'anv': '',
              'id': 3701,
              'join': '',
              'name': 'Crossfire',
              'resource_url': 'https://api.discogs.com/artists/3701',
              'role': '',
              'thumbnail_url': 'https://i.discogs.com/F7Z4BCbnKCkFCls871e6X5DnxO7zpnUX760MlkmKi7Q/rs:fit/g:sm/q:40/h:222/w:600/czM6Ly9kaXNjb2dz/LWRhdGFiYXNlLWlt/YWdlcy9BLTM3MDEt/MTQ4NzI1MzU5Ni0z/NTcwLmpwZWc.jpeg',
              'tracks': ''}],
 'artists_sort': 'Crossfire',
 'barcode': ['5 023903 216700',
             '1210 006 A.',
             '1210 006 AA. NEIL MASTERPIECE '],
 'blocked_from_sale': False,
 'catno': '1210-006',
 'community': {'contributors': [{'resource_url': 'https://api.discogs.com/users/milo',
                                 'username': 'milo'},
                                {'resource_url': 'https://api.discogs.com/users/The_Undertaker',
                                 'username': 'The_Undertaker'},
                                {'resource_url': 'https://api.discogs.com/users/Dandruff',
                                 'username': 'Dandruff'},
                                {'resource_url': 'https://api.discogs.com/users/Aphex139',
                                 'username': 'Aphex139'},
                                {'resource_url': 'https://api.discogs.com/users/the_electrician',
                                 'username': 'the_electrician'},
                                {'resource_url': 'https://api.discogs.com/users/Zubeldia',
                                 'username': 'Zubeldia'},
                                {'resource_url': 'https://api.discogs.com/users/pano9000',
                                 'username': 'pano9000'},
                                {'resource_url': 'https://api.discogs.com/users/jmills',
                                 'username': 'jmills'},
                                {'resource_url': 'https://api.discogs.com/users/mhoys',
                                 'username': 'mhoys'},
                                {'resource_url': 'https://api.discogs.com/users/OllieDee',
                                 'username': 'OllieDee'},
                                {'resource_url': 'https://api.discogs.com/users/vinylmaddy',
                                 'username': 'vinylmaddy'}],
               'data_quality': 'Needs Vote',
               'have': 360,
               'rating': {'average': 3.86, 'count': 66},
               'status': 'Accepted',
               'submitter': {'resource_url': 'https://api.discogs.com/users/milo',
                             'username': 'milo'},
               'want': 80},
 'companies': [{'catno': '',
                'entity_type': '30',
                'entity_type_name': 'Lacquer Cut At',
                'id': 120735,
                'name': 'Masterpiece',
                'resource_url': 'https://api.discogs.com/labels/120735',
                'thumbnail_url': 'https://i.discogs.com/D4k3gvHSqgSNPVXGJ96T5_4O5Y5ep_HQ7TsGpDAWxOI/rs:fit/g:sm/q:40/h:600/w:600/czM6Ly9kaXNjb2dz/LWRhdGFiYXNlLWlt/YWdlcy9MLTEyMDcz/NS0xNTI2MzEyODQ3/LTg4ODQuanBlZw.jpeg'},
               {'catno': '',
                'entity_type': '9',
                'entity_type_name': 'Distributed By',
                'id': 306370,
                'name': 'Vinyl UK',
                'resource_url': 'https://api.discogs.com/labels/306370',
                'thumbnail_url': 'https://i.discogs.com/oKWJ70-YpIGeM8_24PvCJuHyTdRlW5vPxoMvnNcGMAc/rs:fit/g:sm/q:40/h:150/w:150/czM6Ly9kaXNjb2dz/LWRhdGFiYXNlLWlt/YWdlcy9MLTMwNjM3/MC0xNTU1MzUyMTI5/LTQ4ODkuanBlZw.jpeg'},
               {'catno': '',
                'entity_type': '21',
                'entity_type_name': 'Published By',
                'id': 48207,
                'name': 'Copyright Control',
                'resource_url': 'https://api.discogs.com/labels/48207',
                'thumbnail_url': 'https://i.discogs.com/FRvFvNsk726fp6xh-pQzN18g_vPp3VwBfgjJEZY3SSg/rs:fit/g:sm/q:40/h:150/w:150/czM6Ly9kaXNjb2dz/LWRhdGFiYXNlLWlt/YWdlcy9MLTQ4MjA3/LTEzNTY5MDQ0ODct/MTkzNy5qcGVn.jpeg'}],
 'country': 'UK',
 'cover_image': 'https://i.discogs.com/p0m8SdTvQcRL7PdbbYxLsqhhG8SM0w1DyQPFn0u9VRg/rs:fit/g:sm/q:90/h:406/w:400/czM6Ly9kaXNjb2dz/LWRhdGFiYXNlLWlt/YWdlcy9SLTc2MzU3/LTExNDE5Mzk3Njgu/anBlZw.jpeg',
 'data_quality': 'Needs Vote',
 'date_added': '2002-11-28T07:40:02-08:00',
 'date_changed': '2022-04-27T21:48:57-07:00',
 'estimated_weight': 230,
 'extraartists': [{'anv': 'Neil',
                   'id': 342597,
                   'join': '',
                   'name': 'Neil Devine',
                   'resource_url': 'https://api.discogs.com/artists/342597',
                   'role': 'Lacquer Cut By',
                   'thumbnail_url': 'https://i.discogs.com/8ffMnvAUQrUFEKPvI77cKuH_OMlz8on4XDp30xtmCqI/rs:fit/g:sm/q:40/h:200/w:200/czM6Ly9kaXNjb2dz/LWRhdGFiYXNlLWlt/YWdlcy9BLTM0MjU5/Ny0xNTI2MzEyMzA5/LTQzMDMuanBlZw.jpeg',
                   'tracks': ''},
                  {'anv': 'D. Van Simaeys',
                   'id': 692992,
                   'join': '',
                   'name': 'Dimitry Van Simaeys',
                   'resource_url': 'https://api.discogs.com/artists/692992',
                   'role': 'Written-By, Producer',
                   'thumbnail_url': '',
                   'tracks': ''}],
 'format': ['Vinyl', '12"', '45 RPM'],
 'format_quantity': 1,
 'formats': [{'descriptions': ['12"', '45 RPM'], 'name': 'Vinyl', 'qty': '1'}],
 'genre': ['Electronic'],
 'genres': ['Electronic'],
 'id': 76357,
 'identifiers': [{'description': 'Sleeve',
                  'type': 'Barcode',
                  'value': '5 023903 216700'},
                 {'description': 'Side A',
                  'type': 'Matrix / Runout',
                  'value': '1210 006 A.'},
                 {'description': 'Side B',
                  'type': 'Matrix / Runout',
                  'value': '1210 006 AA. NEIL MASTERPIECE '}],
 'images': [{'height': 406,
             'resource_url': 'https://i.discogs.com/p0m8SdTvQcRL7PdbbYxLsqhhG8SM0w1DyQPFn0u9VRg/rs:fit/g:sm/q:90/h:406/w:400/czM6Ly9kaXNjb2dz/LWRhdGFiYXNlLWlt/YWdlcy9SLTc2MzU3/LTExNDE5Mzk3Njgu/anBlZw.jpeg',
             'type': 'primary',
             'uri': 'https://i.discogs.com/p0m8SdTvQcRL7PdbbYxLsqhhG8SM0w1DyQPFn0u9VRg/rs:fit/g:sm/q:90/h:406/w:400/czM6Ly9kaXNjb2dz/LWRhdGFiYXNlLWlt/YWdlcy9SLTc2MzU3/LTExNDE5Mzk3Njgu/anBlZw.jpeg',
             'uri150': 'https://i.discogs.com/lovwaNWoNo7n5_dg7hmp4zCT6PsHFkxltPQm-ChmQgM/rs:fit/g:sm/q:40/h:150/w:150/czM6Ly9kaXNjb2dz/LWRhdGFiYXNlLWlt/YWdlcy9SLTc2MzU3/LTExNDE5Mzk3Njgu/anBlZw.jpeg',
             'width': 400},
            {'height': 406,
             'resource_url': 'https://i.discogs.com/5RhBGHon8hAhOHOJRjWP-ZpXAM0l5BTYIOFMhAeXWR8/rs:fit/g:sm/q:90/h:406/w:400/czM6Ly9kaXNjb2dz/LWRhdGFiYXNlLWlt/YWdlcy9SLTc2MzU3/LTExNDE5Mzk3Nzku/anBlZw.jpeg',
             'type': 'secondary',
             'uri': 'https://i.discogs.com/5RhBGHon8hAhOHOJRjWP-ZpXAM0l5BTYIOFMhAeXWR8/rs:fit/g:sm/q:90/h:406/w:400/czM6Ly9kaXNjb2dz/LWRhdGFiYXNlLWlt/YWdlcy9SLTc2MzU3/LTExNDE5Mzk3Nzku/anBlZw.jpeg',
             'uri150': 'https://i.discogs.com/XV2hH6p-hwa3WkeuoNZT-Vv5sjQxybQumErK94TC9O8/rs:fit/g:sm/q:40/h:150/w:150/czM6Ly9kaXNjb2dz/LWRhdGFiYXNlLWlt/YWdlcy9SLTc2MzU3/LTExNDE5Mzk3Nzku/anBlZw.jpeg',
             'width': 400},
            {'height': 598,
             'resource_url': 'https://i.discogs.com/LMSfQzLgS1vEmkvVG7mqzTrK6Dwply5HCTt1CMcY_bE/rs:fit/g:sm/q:90/h:598/w:600/czM6Ly9kaXNjb2dz/LWRhdGFiYXNlLWlt/YWdlcy9SLTc2MzU3/LTE2MTQ5ODQ5ODEt/NTQ3Ny5qcGVn.jpeg',
             'type': 'secondary',
             'uri': 'https://i.discogs.com/LMSfQzLgS1vEmkvVG7mqzTrK6Dwply5HCTt1CMcY_bE/rs:fit/g:sm/q:90/h:598/w:600/czM6Ly9kaXNjb2dz/LWRhdGFiYXNlLWlt/YWdlcy9SLTc2MzU3/LTE2MTQ5ODQ5ODEt/NTQ3Ny5qcGVn.jpeg',
             'uri150': 'https://i.discogs.com/6HyQ2zEmlCUTI3H658oDhEtAWYKzrOnSNUZhhnpzTNE/rs:fit/g:sm/q:40/h:150/w:150/czM6Ly9kaXNjb2dz/LWRhdGFiYXNlLWlt/YWdlcy9SLTc2MzU3/LTE2MTQ5ODQ5ODEt/NTQ3Ny5qcGVn.jpeg',
             'width': 600},
            {'height': 592,
             'resource_url': 'https://i.discogs.com/g6_3YMeMA2f4bKNZCZDji_36bfJPtl8IgarxhZF72Wo/rs:fit/g:sm/q:90/h:592/w:600/czM6Ly9kaXNjb2dz/LWRhdGFiYXNlLWlt/YWdlcy9SLTc2MzU3/LTE2MTQ5ODQ5OTQt/OTU3NS5qcGVn.jpeg',
             'type': 'secondary',
             'uri': 'https://i.discogs.com/g6_3YMeMA2f4bKNZCZDji_36bfJPtl8IgarxhZF72Wo/rs:fit/g:sm/q:90/h:592/w:600/czM6Ly9kaXNjb2dz/LWRhdGFiYXNlLWlt/YWdlcy9SLTc2MzU3/LTE2MTQ5ODQ5OTQt/OTU3NS5qcGVn.jpeg',
             'uri150': 'https://i.discogs.com/wn5Ticd8KZa3wqj_8eAgjs4gPynxWEcw1MxWxXNdHnM/rs:fit/g:sm/q:40/h:150/w:150/czM6Ly9kaXNjb2dz/LWRhdGFiYXNlLWlt/YWdlcy9SLTc2MzU3/LTE2MTQ5ODQ5OTQt/OTU3NS5qcGVn.jpeg',
             'width': 600}],
 'label': ['1210 Recordings', 'Masterpiece', 'Vinyl UK', 'Copyright Control'],
 'labels': [{'catno': '1210-006',
             'entity_type': '1',
             'entity_type_name': 'Label',
             'id': 182264,
             'name': '1210 Recordings',
             'resource_url': 'https://api.discogs.com/labels/182264',
             'thumbnail_url': 'https://i.discogs.com/nWDUE88YKF6q_BbbBAC1WppiNNvo2Nj_EQpDTyr0BfE/rs:fit/g:sm/q:40/h:418/w:422/czM6Ly9kaXNjb2dz/LWRhdGFiYXNlLWlt/YWdlcy9MLTE4MjI2/NC0xMzcwOTg1MzMw/LTM5NjYuanBlZw.jpeg'}],
 'lowest_price': 2.17,
 'master_id': 225941,
 'master_url': 'https://api.discogs.com/masters/225941',
 'notes': 'Made in England',
 'num_for_sale': 15,
 'released': '2002',
 'released_formatted': '2002',
 'resource_url': 'https://api.discogs.com/releases/76357',
 'series': [],
 'status': 'Accepted',
 'style': ['Drum n Bass'],
 'styles': ['Drum n Bass'],
 'thumb': 'https://i.discogs.com/lovwaNWoNo7n5_dg7hmp4zCT6PsHFkxltPQm-ChmQgM/rs:fit/g:sm/q:40/h:150/w:150/czM6Ly9kaXNjb2dz/LWRhdGFiYXNlLWlt/YWdlcy9SLTc2MzU3/LTExNDE5Mzk3Njgu/anBlZw.jpeg',
 'title': 'Hideout / Swordfight',
 'tracklist': [{'duration': '6:48',
                'position': 'A',
                'title': 'Hideout',
                'type_': 'track'},
               {'duration': '6:22',
                'position': 'B',
                'title': 'Swordfight',
                'type_': 'track'}],
 'type': 'release',
 'uri': 'https://www.discogs.com/release/76357-Crossfire-Hideout-Swordfight',
 'user_data': {'in_collection': False, 'in_wantlist': False},
 'videos': [{'description': 'Official dj Crossfire page   »  '
                            'www.facebook.com/crossfirednb   »   '
                            'crossfire.bandcamp.com   »   '
                            'www.soundcloud.com/crossfire',
             'duration': 416,
             'embed': True,
             'title': 'Crossfire - Hideout (Original Mix)',
             'uri': 'https://www.youtube.com/watch?v=kIZll3BmDI4'},
            {'description': 'Official dj Crossfire page   »  '
                            'www.facebook.com/crossfirednb   »   '
                            'crossfire.bandcamp.com   »   '
                            'www.soundcloud.com/crossfire',
             'duration': 391,
             'embed': True,
             'title': 'Crossfire - Swordfight (Original Mix)',
             'uri': 'https://www.youtube.com/watch?v=mHc_i6AhIZk'}],
 'year': 2002}
```

Reply:
Ah! Sorry! I should have read your 2 latest commits where you actually adress this and make sure data_url from album_info is applied to each track! Nice!